AI scribe UI

zatteo opened the pull request:
Big Scribe PR https://github.com/linagora/tmail-flutter/pull/4187 has been split in two:
- logical part https://github.com/linagora/tmail-flutter/pull/4196 (merged in the features/ai branch)
- UI part (here)

## About prompts

A small PR will come soon that slighty update them to make them work as good as we can quickly. But later, we will handle differently prompts: 
- they will be fetched from a remote location
- they will be managed by a dedicated library or SDK 

## Video

https://github.com/user-attachments/assets/43b869a2-b9bc-47b8-82ea-08cb120f9960



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added AI Scribe integration enabling users to generate text suggestions directly in the composer via AI-powered assistance.
  * Introduced AI assistant button in composer with menu for predefined actions and custom prompts.
  * Added multilingual support for AI assistant across French, Russian, and Vietnamese.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

zatteo committed:
Implement text selection mixin in mobile and web editor

zatteo committed:
Forward text selection callback from editor view to editor

zatteo committed:
Open AI Scribe from bottom bar button

zatteo committed:
Open AI Scribe from text selection

zatteo committed:
Extract method to convert html content to text content

Editor exposes html content (so with html tags) and we do not want to send them to AI backend

zatteo committed:
Use bindings to better separate scribe presentation and data layer

zatteo committed:
Do not display AI scribe if AI is not enabled in env.file

zatteo committed:
Add AI scribe widgets

- Button : the button with sparkle icon to display AI scribe
- Menu : the menu where you can select the action (eg: translate)
- Bar : the input where you can write a custom action
- Suggestion : the wrapper of the result of the LLM

zatteo committed:
Add methods to display AI scribe

dab246 commented:
<img width="648" height="308" alt="Screenshot 2025-12-11 at 13 13 41" src="https://github.com/user-attachments/assets/77763d51-aab9-4ef5-80a2-3769c501b93a" />

- [ ] IMO, We should show submenus when hovering over an item, instead of clicking, so users know which item is selected and which submenu it belongs to.

dab246 commented on lib/features/composer/presentation/widgets/mobile/tablet_bottom_bar_composer_widget.dart:
```diff
@@ -71,6 +75,18 @@ class TabletBottomBarComposerWidget extends StatelessWidget {
+              iconSize: TabletBottomBarComposerWidgetStyle.iconSize,
+              tooltipMessage: AppLocalizations.of(context).aiAssistant,
+              onTapActionCallback: onOpenAIScribe!,
+            ),
```
We should create a separate widget for it so we can reuse it in multiple places.

dab246 commented on lib/features/composer/presentation/widgets/web/web_editor_widget.dart:
```diff
@@ -101,6 +111,19 @@ class _WebEditorState extends State<WebEditorWidget> {
+        if (data['name'] == HtmlUtils.registerSelectionChangeListener.name) {
+          handleSelectionChange(data);
+        }
+      }
```
Should use try/catch for this

dab246 commented on lib/features/composer/presentation/widgets/web/web_editor_widget.dart:
```diff
@@ -101,6 +111,19 @@ class _WebEditorState extends State<WebEditorWidget> {
+      }
+    };
+    if (_selectionChangeListener != null) {
+      window.addEventListener("message", _selectionChangeListener!);
```
We should listen for the `message` event in one place to avoid calling it multiple times. We could reuse `window.addEventListener("message", _dropListener!);`, but it should be modified appropriately for general use, as we might have multiple features that need to be handled in this function.

dab246 commented on lib/features/composer/presentation/composer_controller.dart:
```diff
@@ -859,6 +869,100 @@ class ComposerController extends BaseController
+        .replaceAll('&#39;', "'");
+
+      return textContent.trim();
+  }
```
It should be placed in the util class within the core module for reuse.

dab246 commented on lib/features/composer/presentation/composer_view.dart:
```diff
@@ -413,28 +425,38 @@ class ComposerView extends GetWidget<ComposerController> {
+                            ],
+                          ),
+                          _buildAIScribeSelectionButton(context),
+                        ],
```
Should only add the Stack widget and _buildAIScribeSelectionButton when AIScribe is enabled and AI capability is supported.

zatteo replied:
First step done here https://github.com/linagora/tmail-flutter/pull/4197/changes/75758eb93487f840d5cff4fcd1975c0b36cfbf60 and I plan to add the AI capability in a next PR

dab246 commented on scribe/lib/scribe/ai/presentation/widgets/ai_scribe_bar.dart:
```diff
@@ -0,0 +1,89 @@
+  void _onTextChanged() {
+    setState(() {
+      _isButtonEnabled = _controller.text.trim().isNotEmpty;
+    });
```
If only a small widget is being changed, use `ValueNotifier` instead of `setState`.

dab246 commented on scribe/lib/scribe/ai/presentation/widgets/ai_scribe_menu.dart:
```diff
@@ -0,0 +1,250 @@
+      crossAxisAlignment: CrossAxisAlignment.stretch,
+      children: categories.map((category) {
+        return _buildCategoryItem(category);
+      }).toList(),
```
```suggestion
children: categories.map(_buildCategoryItem).toList(),
```

dab246 commented on scribe/lib/scribe/ai/presentation/widgets/ai_scribe_menu.dart:
```diff
@@ -0,0 +1,250 @@
+    if (category.hasSubmenu) {
+      return Container(
+        key: _categoryKeys[category],
+        child: _buildMenuItem(
```
Why not just pass the key directly to _buildMenuItem?

zatteo replied:
Actually it has no purpose anymore and I forgot to remove it during refactoring and commit merging. I removed it.

zatteo committed:
Show AI Scribe submenu on hover instead of on click

zatteo committed:
fixup! Implement text selection mixin in mobile and web editor

Add try/catch to JSON parse

zatteo committed:
Mutualize event listener in web editor widget

Instead of listening multiple times for message from editor JS side,
let's add only one listener and route inside the callback.

zatteo committed:
Simplify observed values about text selection in composer

hasTextSelection had no interest.

The other two values have been put inside an object to observe only one object.

zatteo committed:
fixup! Extract method to convert html content to text content

Move the method in utils

zatteo committed:
Export AI Scribe code from controller to a mixin

Avoid adding too much code in the controller

zatteo commented on scribe/lib/scribe/ai/presentation/widgets/ai_scribe.dart:
```diff
@@ -0,0 +1,213 @@
+  );
+}
+
+({double left, double bottom}) _calculateModalPosition({
```
Good idea. It will also help me to fix some edge cases I saw this morning.

zatteo committed:
Remove unused categoryKeys

zatteo committed:
Use ValueNotifier instead of setState in AIScribeBar

zatteo committed:
Add tests to AI scribe calculcateModalPosition

zatteo committed:
Fix AI scribe menu going off screen to the top

If it was opened at the top of the composer, it could go off screen to the top.

modalHeight was missing so the calculateModalPosition method could not protect the AI scribe menu to go off screen from the top.

zatteo committed:
Fix tests by initializing DotEnv for testing

Crash-- commented:
> The way we generally move forward is by leveraging a JMAP capability to advertize back support.


@chibenwa next PR https://github.com/linagora/tmail-flutter/pull/4203

chibenwa commented on core/lib/utils/string_convert.dart:
```diff
@@ -196,4 +196,18 @@ class StringConvert {
+      .replaceAll('&gt;', '>')
+      .replaceAll('&quot;', '"')
+      .replaceAll('&#39;', "'");
+
```
> What do you think of this piece of Flutter code? Should I roast it?

# 🔍 **What’s good**

* The function is simple and readable.
* For very basic HTML, it might work.

But…

---

# ⚠️ **The issues (yes, you can roast this code 😄)**

## 1. **Using a regex to strip HTML tags…**

This is *always* fragile.

Problems you will eventually hit:

* Tags with attributes containing `>`
* Comments, scripts, styles
* Broken or unusual HTML
* Nested content where regex fails

👉 **There’s a reason HTML parsers exist.**

---

## 2. **Only a few HTML entities are handled**

There are **thousands** of HTML entities.
Your code handles maybe 5.

Anything like `&eacute;`, `&copy;`, `&hellip;`, etc. will remain unchanged.

---

## 3. **No handling of `<br>`, `<p>`, etc.**

All line breaks will disappear.
Your output becomes a single messy block:

> `"Line1Line2Line3"`

---

## 4. **Whitespace is not normalized**

HTML often contains multiple spaces, newlines, indentation.
Your output might be weirdly spaced.

---

# ✔️ **A proper solution in Flutter**

### Use the `html` package:

```dart
import 'package:html/parser.dart' as html;

String convertHtmlContentToTextContent(String htmlContent) {
  final document = html.parse(htmlContent);
  final String parsedString = document.body?.text ?? '';
  return parsedString.trim();
}
```

### Why this is better:

* Correctly strips tags
* Correctly decodes *all* HTML entities (`&eacute;` → `é`)
* More robust
* Easier to maintain


---

# 🧠 **Conclusion**

Your current code works **only for very simple HTML** and is **not reliable** for real-world HTML content.

👉 Strong recommendation: use an HTML parser instead.

chibenwa replied:
And yes, I fucking hate REGEXPs...

chibenwa commented on core/lib/utils/string_convert.dart:
```diff
@@ -196,4 +196,18 @@ class StringConvert {
+      .replaceAll('&gt;', '>')
+      .replaceAll('&quot;', '"')
+      .replaceAll('&#39;', "'");
+
```
> Redos vulnerable?

Short answer: **Yes — this regex *can* be ReDoS-vulnerable depending on the input.**

Let’s break it down clearly.

---

# ✅ **Your regex**

```dart
RegExp(r'<[^>]*>')
```

# ⚠️ **Is it vulnerable to ReDoS?**

**Potentially, yes — in specific edge cases.**

Reason:

* `<[^>]*>` uses `[^>]` with `*`, which is a **greedy, unbounded quantifier**.
* On malformed or adversarial input designed to confuse the engine, it can trigger **catastrophic backtracking**.

### Example of problematic input

A long string like:

```
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<!
```

Or nested angled brackets like:

```
< < < < < < < < < < < < < < x
```

The regex tries to match:

* The opening `<`
* Then tries to find the next `>`, but finds none
* It backtracks repeatedly through a huge number of combinations

This can cause:

* CPU spike
* App freeze
* DoS in worst cases

---

# 🔍 **Why this happens**

The `[^>]` is fine by itself, but when used like this:

```
< [^>]* >
```

…it forces the engine to try **every possible way** to match the sequence when no matching `>` exists.

This is a textbook case of a **super-linear** regex pattern.

---

# ✔️ **How to avoid ReDoS**

## 1. Use a *non-greedy* version

```dart
RegExp(r'<[^>]*?>');
```

Still fragile, but **less prone** to catastrophic backtracking.

---

## 2. Use a *possessive* quantifier (not available in Dart RegExp 😞)

In languages like Java:

```
<[^>]*+>
```

Dart does **not** support this — so you can’t fully fix it with regex alone.

---

## 3. **Best solution: don’t use regex for HTML parsing**

Dart has an HTML parser that **avoids regex pitfalls** entirely:

```dart
import 'package:html/parser.dart' as html;

String convertHtmlContentToTextContent(String htmlContent) {
  final document = html.parse(htmlContent);
  return document.body?.text.trim() ?? '';
}
```

This eliminates:

* ReDoS risk
* Invalid HTML issues
* Missed entities
* Missing line breaks

---

# 🧠 **Conclusion**

**Yes**, your regex can be ReDoS-vulnerable with adversarial input.

If this code runs on:

* user-generated content
* emails
* scraped HTML
* chat messages

👉 **Do NOT use regex. Use an HTML parser.**

If you want, I can also analyze your *specific* use-case and tell you the actual risk level.

chibenwa replied:
(Or use a re2 engine without backtracking if really forced with a regexp)

zatteo replied:
Very good point. I proposed a new version here https://github.com/linagora/tmail-flutter/pull/4197/changes/5f3b46dedacb59b7cad87dc40e907a053c8e25c6. FYI in the future (but I don't know if we talk about days or weeks or more), we may need to keep HTML here to be able to keep the styling. But for now it is simpler like this as it avoid to render HTML in the scribe.

chibenwa replied:
MUCH better indeed. Thanks.

zatteo committed:
Use html parser instead of regex to clean composer input

Before sending input to the LLM, we clean the composer input to
remove html tags. We relied on regex that we not complete enough and
could be bypassed easily. So let's use the html parser available in
Dart instead.

I also added some tests.

chibenwa reviewed (approved):
Still many changesout of lib/scribe but I'm unsure committing to more refactorings here on this specific feature is worth it.

zatteo committed:
fixup! Use html parser instead of regex to clean composer input

Improve efficiency and add more tests in convertHtmlContentToTextContent

zatteo commented on lib/features/composer/presentation/widgets/web/bottom_bar_composer_widget.dart:
```diff
@@ -31,6 +31,8 @@ class BottomBarComposerWidget extends StatelessWidget {
   final VoidCallback saveAsTemplateAction;
   final VoidCallback onOpenInsertLink;
+  final VoidCallback? onOpenAIScribe;
+  final GlobalKey? aiScribeButtonKey;
```
Understood. WDYT about this totally different approach that removes the aiScribeButtonKey ? https://github.com/linagora/tmail-flutter/pull/4207

hoangdat commented:
All things will be addressed in #4203
